### ISSUE TITLE
code phí bảo hiểm

### ISSUE THREAD

**Issue by thongnguyen1804** [thongnguyen1804/code-ph-b-o-hi-m#1]
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from keras.models import Sequential
from keras.layers import Dense

# ===== 1. Đọc dữ liệu =====
data = pd.read_csv("f:\AI\baohiem.csv.csv")
print("5 dòng đầu dữ liệu:")
print(data.head())

# ===== 2. Tách X và y =====
X = data.drop("charges", axis=1)
y = data["charges"]

# ===== 3. Tiền xử lý =====
numeric_features = ["age", "bmi", "children"]
categorical_features = ["sex", "smoker", "region"]

numeric_transformer = StandardScaler()
categorical_transformer = OneHotEncoder(drop="first")  # tránh đa cộng tuyến

preprocessor = ColumnTransformer(
    transformers=[
        ("num", numeric_transformer, numeric_features),
        ("cat", categorical_transformer, categorical_features)
    ]
)

X_processed = preprocessor.fit_transform(X)

# Chia train/test
X_train, X_test, y_train, y_test = train_test_split(
    X_processed, y, test_size=0.2, random_state=42
)

# ===== 4. Xây dựng ANN =====
model = Sequential()
model.add(Dense(64, activation='relu', input_shape=(X_train.shape[1],)))
model.add(Dense(32, activation='relu'))
model.add(Dense(1, activation='linear'))  # đầu ra liên tục (charges)

# ===== 5. Compile & Train =====
model.compile(optimizer='adam', loss='mse', metrics=['mae'])

history = model.fit(
    X_train, y_train,
    validation_split=0.1,
    epochs=200,
    batch_size=32,
    verbose=1
)
# ===== 6. Đánh giá =====
test_loss, test_mae = model.evaluate(X_test, y_test, verbose=0)
print("Test MAE (Mean Absolute Error):", test_mae)  

# ===== 7. Đánh giá =====
y_pred = model.predict(X_test)
r2 = r2_score(y_test, y_pred)
print("R² score (giống accuracy trong regression):", r2)
